ldap/servers/plugins/roles/roles_cache.c

```diff
@@ -2131,3 +2131,139 @@ roles_cache_dump(caddr_t data, caddr_t arg __attribute__((unused)))
 
     return 0;
 }
+
+
+/* This is an example callback to substitute
+ * attribute type 'from' with 'to' in all
+ * the filter components
+ * example_substitute_type is a callback (FILTER_APPLY_FN) used by slapi_filter_apply
+ * typedef int (FILTER_APPLY_FN)(Slapi_Filter f, void *arg)
+ * To stick to the definition, the callback is defined using 'int' rather 'int32_t'
+ */
+typedef struct {
+    char *attrtype_from;
+    char *attrtype_to;
+} role_substitute_type_arg_t;
+
+
+static void
+_rewrite_nsrole_component(Slapi_Filter *f, role_substitute_type_arg_t *substitute_arg)
+{
+    char *type;
+    struct berval *bval;
+    char *attrs[3] = {SLAPI_ATTR_OBJECTCLASS, ROLE_FILTER_ATTR_NAME, NULL};
+    Slapi_Entry *nsrole_entry = NULL;
+    Slapi_DN *sdn = NULL;
+    char *rolefilter = NULL;
+    int rc;
+    char **oc_values = NULL;
+
+    /* Substitution is only valid if original attribute is NSROLEATTR */
+    if ((substitute_arg == NULL) ||
+        (substitute_arg->attrtype_from == NULL) ||
+        (substitute_arg->attrtype_to == NULL)) {
+        return;
+    }
+    if (strcasecmp(substitute_arg->attrtype_from, NSROLEATTR)) {
+        return;
+    }
+
+    /* Check that assertion does not refer to a filter/nested role */
+    if (slapi_filter_get_ava(f, &type, &bval)) {
+        return;
+    }
+    sdn = slapi_sdn_new_dn_byref(bval->bv_val);
+    if (slapi_search_internal_get_entry(sdn, attrs, &nsrole_entry, roles_get_plugin_identity())) {
+        goto bail;
```

Comment:
Could probably do something much clever in this case:
if the role does not exists then the filter does not match anything. 
 (goto bail means having an unindexed search)

Reply:
I did the change so that a component with a not existing role will be replaced by an indexed component with IDL=0. "(nsuniqueid=-1)"